proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,376 @@
+
+Existing work on higher order unification has revealed that higher order unification of the template application form is a useful decidable subset of the general problem. This was first observed by Miller in `"Unification Under a Mixed Prefix" <https://repository.upenn.edu/cis_reports/454/>`_. Miller called these template applications 'patterns', but we decided against using that terminology because of the confusion with regular pattern matching in Haskell. Nowadays higher unification with Miller's pattern restriction is commonplace in dependently typed languages such as `Agda <http://www.cse.chalmers.se/~ulfn/papers/thesis.pdf>`_ and `Idris <https://www.type-driven.org.uk/edwinb/papers/impldtp.pdf>`_, which shows that higher order unification with this restriction is still very powerful.
+
+One notable extension of Miller's patterns is the extension by Duggan in his paper `"Unification with extended patterns" <https://doi.org/10.1016/S0304-3975(97)00141-2>`_. Duggan extends the simple template application form to allow projections of local binders as arguments to the template variables. Such an extension might also be applicable to the template application form in this proposal.
+
+Examples
+--------
+
+* 	One of the simplest examples is this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x) = "RULE FIRED" #-}
+
+	It would match expressions like:
+	::
+
+		foo (\x -> x * 2 + x)
+
+* 	The higher order pattern may involve multiple locally bound variables, e.g.:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y z -> f x y z) = "RULE FIRED" #-}
+
+	Which would match:
+	::
+
+		foo (\x y z -> x * y + z)
+
+	But not every variable has to occur in the match. It would also match this expression where ``y`` does not occur:
+	::
+
+		foo (\x y z -> x * 2 + z)
+
+* 	Locally bound variables may only occur once.
+	Consider the following rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x x) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x -> x * 2 + x)
+
+	But it does contain the valid subrule ``f x``, so it would match:
+	::
+
+		foo (\x -> (bar x . baz) x)
+
+* 	Similarly if the template variable ``f`` is applied to non-variable arguments then it only matches a literal application.
+	Consider this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y -> f x 2 y) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x y -> x * 2 + y)`
```

Comment:
```suggestion
		foo (\x _ y -> x * 2 + y)`
```

perhaps? otherwise, where does the `2` arg go?

Reply:
No I don't think so.  I don't see the problem here.

Reply:
Oh, right. Sorry for the confusion.